File: scraper/src/jobscraper/skills.py
"""Curated skill dictionary. Each entry: canonical_name -> (category, aliases).

Categories: lang, framework, frontend, db, cloud, devops, data, ai, mobile,
tools, methodology, soft.

Aliases must be lowercase. Matching uses word boundaries on lowercased
description text.
"""
from __future__ import annotations

SKILLS: dict[str, tuple[str, list[str]]] = {
    # Languages
    "Python": ("lang", ["python"]),
    "JavaScript": ("lang", ["javascript", "js"]),
    "TypeScript": ("lang", ["typescript", "ts"]),
    "Java": ("lang", ["java"]),
    "C": ("lang", ["c programming", " c "]),
    "C++": ("lang", ["c\\+\\+", "cpp"]),
    "C#": ("lang", ["c#", "csharp"]),
    "Go": ("lang", ["golang", " go "]),
    "Rust": ("lang", ["rust"]),
    "Ruby": ("lang", ["ruby"]),
    "PHP": ("lang", ["php"]),
    "Kotlin": ("lang", ["kotlin"]),
    "Swift": ("lang", ["swift"]),
    "Scala": ("lang", ["scala"]),
    "SQL": ("lang", ["sql"]),
    "Bash": ("lang", ["bash", "shell scripting"]),
    "R": ("lang", [" r ,"]),

    # Frameworks (backend)
    "FastAPI": ("framework", ["fastapi"]),
    "Django": ("framework", ["django"]),
    "Flask": ("framework", ["flask"]),
    "Node.js": ("framework", ["node\\.js", "nodejs", "node js"]),
    "Express": ("framework", ["express\\.js", "expressjs", "express"]),
    "NestJS": ("framework", ["nestjs", "nest\\.js"]),
    "Spring": ("framework", ["spring boot", "spring"]),
    "Rails": ("framework", ["ruby on rails", "rails"]),
    "Laravel": ("framework", ["laravel"]),
    "ASP.NET": ("framework", ["asp\\.net", "\\.net core", "\\.net"]),
    "GraphQL": ("framework", ["graphql"]),
    "gRPC": ("framework", ["grpc"]),
    "REST": ("framework", ["rest api", "restful"]),

    # Frontend
    "React": ("frontend", ["react\\.js", "reactjs", "react"]),
    "Next.js": ("frontend", ["next\\.js", "nextjs"]),
    "Vue": ("frontend", ["vue\\.js", "vuejs", "vue"]),
    "Nuxt": ("frontend", ["nuxt"]),
    "Angular": ("frontend", ["angular"]),
    "Svelte": ("frontend", ["svelte", "sveltekit"]),
    "HTML": ("frontend", ["html5?"]),
    "CSS": ("frontend", ["css3?"]),
    "Tailwind": ("frontend", ["tailwind"]),
    "Sass": ("frontend", ["sass", "scss"]),
    "Redux": ("frontend", ["redux"]),
    "Vite": ("frontend", ["vite"]),
    "Webpack": ("frontend", ["webpack"]),

    # Databases
    "PostgreSQL": ("db", ["postgresql", "postgres"]),
    "MySQL": ("db", ["mysql", "mariadb"]),
    "SQLite": ("db", ["sqlite"]),
    "MongoDB": ("db", ["mongodb", "mongo"]),
    "Redis": ("db", ["redis"]),
    "Elasticsearch": ("db", ["elasticsearch", "elastic search"]),
    "Cassandra": ("db", ["cassandra"]),
    "DynamoDB": ("db", ["dynamodb"]),
    "SQL Server": ("db", ["sql server", "mssql"]),
    "Oracle DB": ("db", ["oracle db", "oracle database"]),

    # Cloud
    "AWS": ("cloud", ["aws", "amazon web services"]),
    "GCP": ("cloud", ["gcp", "google cloud"]),
    "Azure": ("cloud", ["azure"]),
    "S3": ("cloud", ["s3"]),
    "EC2": ("cloud", ["ec2"]),
    "Lambda": ("cloud", ["aws lambda", "lambda function"]),
    "Cloudflare": ("cloud", ["cloudflare"]),
    "Vercel": ("cloud", ["vercel"]),
    "Heroku": ("cloud", ["heroku"]),
    "DigitalOcean": ("cloud", ["digitalocean", "digital ocean"]),

    # DevOps
    "Docker": ("devops", ["docker"]),
    "Kubernetes": ("devops", ["kubernetes", "k8s"]),
    "Terraform": ("devops", ["terraform"]),
    "Ansible": ("devops", ["ansible"]),
    "Jenkins": ("devops", ["jenkins"]),
    "GitHub Actions": ("devops", ["github actions"]),
    "GitLab CI": ("devops", ["gitlab ci", "gitlab-ci"]),
    "CircleCI": ("devops", ["circleci", "circle ci"]),
    "Linux": ("devops", ["linux"]),
    "Nginx": ("devops", ["nginx"]),
    "Prometheus": ("devops", ["prometheus"]),
    "Grafana": ("devops", ["grafana"]),
    "CI/CD": ("devops", ["ci/cd", "ci / cd", "continuous integration"]),

    # Data / AI
    "Pandas": ("data", ["pandas"]),
    "NumPy": ("data", ["numpy"]),
    "Spark": ("data", ["apache spark", "pyspark", "spark"]),
    "Kafka": ("data", ["kafka"]),
    "Airflow": ("data", ["airflow"]),
    "dbt": ("data", ["dbt"]),
    "Snowflake": ("data", ["snowflake"]),
    "BigQuery": ("data", ["bigquery"]),
    "ETL": ("data", ["etl", "elt"]),
    "TensorFlow": ("ai", ["tensorflow"]),
    "PyTorch": ("ai", ["pytorch"]),
    "scikit-learn": ("ai", ["scikit-learn", "sklearn"]),
    "LangChain": ("ai", ["langchain"]),
    "LLM": ("ai", ["llm", "large language model"]),
    "Machine Learning": ("ai", ["machine learning"]),
    "Deep Learning": ("ai", ["deep learning"]),
    "NLP": ("ai", ["nlp", "natural language processing"]),

    # Mobile
    "React Native": ("mobile", ["react native"]),
    "Flutter": ("mobile", ["flutter"]),
    "Android": ("mobile", ["android"]),
    "iOS": ("mobile", ["ios development", " ios "]),

    # Tools / methodology
    "Git": ("tools", ["git"]),
    "GitHub": ("tools", ["github"]),
    "GitLab": ("tools", ["gitlab"]),
    "Jira": ("tools", ["jira"]),
    "Figma": ("tools", ["figma"]),
    "Agile": ("methodology", ["agile"]),
    "Scrum": ("methodology", ["scrum"]),
    "Kanban": ("methodology", ["kanban"]),
    "TDD": ("methodology", ["tdd", "test-driven", "test driven"]),
    "DDD": ("methodology", ["domain-driven", "ddd"]),
    "Microservices": ("methodology", ["microservices", "micro-services"]),
    "OOP": ("methodology", ["object-oriented", "oop"]),
    "Functional Programming": ("methodology", ["functional programming"]),
    "Design Patterns": ("methodology", ["design patterns"]),
    "Unit Testing": ("methodology", ["unit test", "unit testing"]),

    # Soft
    "English": ("soft", ["english (b2|c1|c2|fluent|proficient|advanced)", "fluent english"]),
    "Communication": ("soft", ["communication skills"]),
    "Teamwork": ("soft", ["teamwork", "team player"]),
    "Problem Solving": ("soft", ["problem[- ]solving"]),
}


SENIORITY_PATTERNS = {
    "intern": [r"\bintern\b", r"\binternship\b", r"\bstajyer\b", r"\bstaj\b"],
    "junior": [r"\bjunior\b", r"\bjr\.?\b", r"\bentry[- ]level\b", r"\bgraduate\b", r"\bnew grad\b", r"0[-+ ]?[12]\s*years?", r"\bjr\b"],
    "mid": [r"\bmid[- ]level\b", r"\bintermediate\b", r"[2-5]\s*[-+]?\s*years?"],
    "senior": [r"\bsenior\b", r"\bsr\.?\b", r"\blead\b", r"\bstaff\b", r"\bprincipal\b", r"[5-9]\+?\s*years?", r"10\+?\s*years?"],
}

File: web/convex/skills.ts
// Curated skill dictionary. Mirrors scraper/jobscraper/skills.py:SKILLS so
// the Convex side can detect market-demanded skills inside job text without
// needing a Python service. Aliases are lowercase; matching is word-bound
// against the lowercased job text.

export type SkillEntry = { canonical: string; category: string; aliases: string[] };

export const SKILLS: SkillEntry[] = [
  // Languages
  { canonical: "Python", category: "lang", aliases: ["python"] },
  { canonical: "JavaScript", category: "lang", aliases: ["javascript", "js"] },
  { canonical: "TypeScript", category: "lang", aliases: ["typescript", "ts"] },
  { canonical: "Java", category: "lang", aliases: ["java"] },
  { canonical: "C++", category: "lang", aliases: ["c\\+\\+", "cpp"] },
  { canonical: "C#", category: "lang", aliases: ["c#", "csharp"] },
  { canonical: "Go", category: "lang", aliases: ["golang"] },
  { canonical: "Rust", category: "lang", aliases: ["rust"] },
  { canonical: "Ruby", category: "lang", aliases: ["ruby"] },
  { canonical: "PHP", category: "lang", aliases: ["php"] },
  { canonical: "Kotlin", category: "lang", aliases: ["kotlin"] },
  { canonical: "Swift", category: "lang", aliases: ["swift"] },
  { canonical: "Scala", category: "lang", aliases: ["scala"] },
  { canonical: "SQL", category: "lang", aliases: ["sql"] },
  { canonical: "Bash", category: "lang", aliases: ["bash", "shell scripting"] },

  // Frameworks (backend)
  { canonical: "FastAPI", category: "framework", aliases: ["fastapi"] },
  { canonical: "Django", category: "framework", aliases: ["django"] },
  { canonical: "Flask", category: "framework", aliases: ["flask"] },
  { canonical: "Node.js", category: "framework", aliases: ["node\\.js", "nodejs", "node js"] },
  { canonical: "Express", category: "framework", aliases: ["express\\.js", "expressjs", "express"] },
  { canonical: "NestJS", category: "framework", aliases: ["nestjs", "nest\\.js"] },
  { canonical: "Spring", category: "framework", aliases: ["spring boot", "spring"] },
  { canonical: "Rails", category: "framework", aliases: ["ruby on rails", "rails"] },
  { canonical: "Laravel", category: "framework", aliases: ["laravel"] },
  { canonical: "ASP.NET", category: "framework", aliases: ["asp\\.net", "\\.net core", "\\.net"] },
  { canonical: "GraphQL", category: "framework", aliases: ["graphql"] },
  { canonical: "gRPC", category: "framework", aliases: ["grpc"] },
  { canonical: "REST", category: "framework", aliases: ["rest api", "restful"] },

  // Frontend
  { canonical: "React", category: "frontend", aliases: ["react\\.js", "reactjs", "react"] },
  { canonical: "Next.js", category: "frontend", aliases: ["next\\.js", "nextjs"] },
  { canonical: "Vue", category: "frontend", aliases: ["vue\\.js", "vuejs", "vue"] },
  { canonical: "Nuxt", category: "frontend", aliases: ["nuxt"] },
  { canonical: "Angular", category: "frontend", aliases: ["angular"] },
  { canonical: "Svelte", category: "frontend", aliases: ["svelte", "sveltekit"] },
  { canonical: "HTML", category: "frontend", aliases: ["html5?"] },
  { canonical: "CSS", category: "frontend", aliases: ["css3?"] },
  { canonical: "Tailwind", category: "frontend", aliases: ["tailwind"] },
  { canonical: "Sass", category: "frontend", aliases: ["sass", "scss"] },
  { canonical: "Redux", category: "frontend", aliases: ["redux"] },
  { canonical: "Vite", category: "frontend", aliases: ["vite"] },
  { canonical: "Webpack", category: "frontend", aliases: ["webpack"] },

  // Databases
  { canonical: "PostgreSQL", category: "db", aliases: ["postgresql", "postgres"] },
  { canonical: "MySQL", category: "db", aliases: ["mysql", "mariadb"] },
  { canonical: "SQLite", category: "db", aliases: ["sqlite"] },
  { canonical: "MongoDB", category: "db", aliases: ["mongodb", "mongo"] },
  { canonical: "Redis", category: "db", aliases: ["redis"] },
  { canonical: "Elasticsearch", category: "db", aliases: ["elasticsearch", "elastic search"] },
  { canonical: "Cassandra", category: "db", aliases: ["cassandra"] },
  { canonical: "DynamoDB", category: "db", aliases: ["dynamodb"] },
  { canonical: "SQL Server", category: "db", aliases: ["sql server", "mssql"] },

  // Cloud
  { canonical: "AWS", category: "cloud", aliases: ["aws", "amazon web services"] },
  { canonical: "GCP", category: "cloud", aliases: ["gcp", "google cloud"] },
  { canonical: "Azure", category: "cloud", aliases: ["azure"] },
  { canonical: "Cloudflare", category: "cloud", aliases: ["cloudflare"] },
  { canonical: "Vercel", category: "cloud", aliases: ["vercel"] },
  { canonical: "Heroku", category: "cloud", aliases: ["heroku"] },

  // DevOps
  { canonical: "Docker", category: "devops", aliases: ["docker"] },
  { canonical: "Kubernetes", category: "devops", aliases: ["kubernetes", "k8s"] },
  { canonical: "Terraform", category: "devops", aliases: ["terraform"] },
  { canonical: "Ansible", category: "devops", aliases: ["ansible"] },
  { canonical: "Jenkins", category: "devops", aliases: ["jenkins"] },
  { canonical: "GitHub Actions", category: "devops", aliases: ["github actions"] },
  { canonical: "GitLab CI", category: "devops", aliases: ["gitlab ci", "gitlab-ci"] },
  { canonical: "Linux", category: "devops", aliases: ["linux"] },
  { canonical: "Nginx", category: "devops", aliases: ["nginx"] },
  { canonical: "CI/CD", category: "devops", aliases: ["ci/cd", "continuous integration"] },

  // Data / AI
  { canonical: "Pandas", category: "data", aliases: ["pandas"] },
  { canonical: "NumPy", category: "data", aliases: ["numpy"] },
  { canonical: "Spark", category: "data", aliases: ["apache spark", "pyspark", "spark"] },
  { canonical: "Kafka", category: "data", aliases: ["kafka"] },
  { canonical: "Airflow", category: "data", aliases: ["airflow"] },
  { canonical: "Snowflake", category: "data", aliases: ["snowflake"] },
  { canonical: "BigQuery", category: "data", aliases: ["bigquery"] },
  { canonical: "ETL", category: "data", aliases: ["etl", "elt"] },
  { canonical: "TensorFlow", category: "ai", aliases: ["tensorflow"] },
  { canonical: "PyTorch", category: "ai", aliases: ["pytorch"] },
  { canonical: "scikit-learn", category: "ai", aliases: ["scikit-learn", "sklearn"] },
  { canonical: "LangChain", category: "ai", aliases: ["langchain"] },
  { canonical: "Machine Learning", category: "ai", aliases: ["machine learning"] },
  { canonical: "Deep Learning", category: "ai", aliases: ["deep learning"] },
  { canonical: "NLP", category: "ai", aliases: ["nlp", "natural language processing"] },

  // Mobile
  { canonical: "React Native", category: "mobile", aliases: ["react native"] },
  { canonical: "Flutter", category: "mobile", aliases: ["flutter"] },
  { canonical: "Android", category: "mobile", aliases: ["android"] },

  // Tools / methodology
  { canonical: "Git", category: "tools", aliases: ["git"] },
  { canonical: "GitHub", category: "tools", aliases: ["github"] },
  { canonical: "GitLab", category: "tools", aliases: ["gitlab"] },
  { canonical: "Jira", category: "tools", aliases: ["jira"] },
  { canonical: "Figma", category: "tools", aliases: ["figma"] },
  { canonical: "Agile", category: "methodology", aliases: ["agile"] },
  { canonical: "Scrum", category: "methodology", aliases: ["scrum"] },
  { canonical: "TDD", category: "methodology", aliases: ["tdd", "test-driven", "test driven"] },
  { canonical: "Microservices", category: "methodology", aliases: ["microservices", "micro-services"] },
];

// Pre-compile regexes once. Word-bound on lowercased text.
const COMPILED: Array<{ canonical: string; category: string; pattern: RegExp }> = SKILLS.map((s) => ({
  canonical: s.canonical,
  category: s.category,
  pattern: new RegExp(`(?<![a-z0-9_])(?:${s.aliases.join("|")})(?![a-z0-9_])`, "i"),
}));

export function extractSkills(text: string): Array<{ skill: string; category: string }> {
  if (!text) return [];
  const out: Array<{ skill: string; category: string }> = [];
  for (const { canonical, category, pattern } of COMPILED) {
    if (pattern.test(text)) out.push({ skill: canonical, category });
  }
  return out;
}
